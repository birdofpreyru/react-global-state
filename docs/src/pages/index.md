import TsNote from '../../docs/api/_ts-note.md';

# React Global State

[![Latest NPM Release](https://img.shields.io/npm/v/@dr.pogodin/react-global-state.svg)](https://www.npmjs.com/package/@dr.pogodin/react-global-state)
[![NPM monthly downloads](https://img.shields.io/npm/dm/@dr.pogodin/react-global-state)](https://www.npmjs.com/package/@dr.pogodin/react-global-state)
[![CircleCI](https://dl.circleci.com/status-badge/img/gh/birdofpreyru/react-global-state/tree/master.svg?style=shield)](https://app.circleci.com/pipelines/github/birdofpreyru/react-global-state)
[![GitHub Repo stars](https://img.shields.io/github/stars/birdofpreyru/react-global-state?style=social)](https://github.com/birdofpreyru/react-global-state)

**React Global State** is a state of the art library for the global state and
asynchronous data management in [React] applications, powered by [hooks] and
[Context API]. It is simple, efficient, and features a full SSR (server-side
rendering) support.

[![Sponsor](../../../.README/sponsor.svg)](https://github.com/sponsors/birdofpreyru)

<TsNote />

## Motivation (JavaScript)

_**TypeScript**-specific nuances are explained in the next section._

The motivation and vision behind this project is to bring to the table all
useful features of [Redux], without related development overheads, like
the amount of necessary boilerplate code, and the efforts needed to design
and maintain actions and reducers.

With this library, the introduction of a datum (data piece), shared across
different application components, is as easy as using the local React state:


```jsx
function SampleReactComponent() {
  const [data, setData] = useGlobalState('data.storage.path', initialValue);

  /* `data` value can be updating by calling `setData(newData)` anywhere inside
   * the component code, including inside hooks like `useEffect(..)` or some
   * event handlers. */

  return /* Some JSX markup. */;
}
```

Relying on async data, _e.g._ loading into the state data from a 3-rd party API,
is the same easy:

```jsx
async function loader() {

  /* Some async operation to get data, like a call to a 3-rd party API. */

  return data;
}

function SampleReactComponent() {
  const { data, loading, timestamp } = useAsyncData('data.envelope.path', loader);

  /* `data` holds the data loaded into the global state, if they are fresh enough;
   * `loading` signals that data loading (or silent re-loading) is in-progress;
   * `timestamp` is the timestamp of currently loaded `data`. */

  return /* Some JSX markup. */
}
```
&uArr; _Behind the scene, the library takes care about updating the component
when the data loading starts and ends, also about the timestamps, automatic
reloading, and garbage collection of aged data._

Related closely to async data is the server-side rendering (SSR). This library
takes it into account, and provides a flexible way to implement SSR with loading
of some, or all async data at the server side.

## Motivation (TypeScript)

As many real-world projects nowadays are developed in TypeScript (justified so,
or for all wrong reasons), a&nbsp;good JavaScript library has support TypeScript,
and take a full advantage of its features, when available. It is worth reminding,
TypeScript does _static, compile-time type-checks_, thus we cannot, and are not
aiming to use its features for an exhaustive global state validation, casting,
_etc._, which would require dedicated runtime features, and goes beyond
the (current) scope of our library. We rather have the following aims with our
TypeScript support:

- In cases when static typing with TypeScript is able to evaluate the exact type
  of value on a given global state path, we want to use it for type checks.

- In cases when TypeScript is not able to evaluate, or is wrong about the exact
  type of value on a given global state path, we want to be able to force it to
  assume the type we need, whether it is correct or wrong.

- We want to have it very clear from the code, whether TypeScript's type
  resolution checked, and verified a value type, or wether we forced it under
  our own responsibility.

Regarding the evaluation of exact value type on a given global state path,
TypeScript is able to do it when the path string type can be reduced to
a _string literal type_, that describes a valid global state path,
and the type of global state is well-defined along that path. For example:

```ts
import RGS, { type API, type ForceT} from '@dr.pogodin/react-global-state';

type StateT = { some: { path: string }};

// This call returns a set of React components and hooks "locked-in" into
// the same specified state path.
const { useGlobalState } = RGS as API<StateT>;

function SampleReactComponent() {
  // Here TypeScript will automatically resolve "string" as the correct type of
  // `dataA` value and `setDataA` argument (`setDataA` will also accept a correct
  // state update function).
  const [dataA, setDataA] = useGlobalState('some.path');

  // This will report a TypeScript error: "unknown.path" is not a correct path
  // inside our `StateT` type, and thus our library (in TS), by default, rejects
  // to write/read it, although for its runtime implementation it is a legit
  // operation: in the first call to an unknown path, the given initial value,
  // `123`, will be written to that state path, then any hook related to that
  // path will work as usual. This default TS behavior is on purpose, to make
  // its clear that TS does not know given path (the type of value on that path),
  // and thus the user must force it, if he is sure about it.
  const dataB = useGlobalState('unknown.path', 123)[0];

  // This variant forces "number" value type on that unknown path.
  const dataC = useGlobalState<ForceT, number>('unknown.path', 123)[0];

  return /* Some JSX markup. */;
}
```

Here is a brief exampe of async data loading with TS-flavour of this library:

```ts
import RGS, {
  type API,
  type AsyncDataEnvelopeT,
  type ForceT,
} from '@dr.pogodin/react-global-state';

// `AsyncDataEnvelopeT<DataT>` type describes a state segment with stores
// data of given type `DataT`, managed by `useAsyncData()` hook(s).
type StateT = { some: { path: AsyncDataEnvelopeT<string> }};

// This call returns a set of React components and hooks "locked-in" into
// the same specified state path.
const { useAsyncData } = RGS as API<StateT>;

async function loader(): string {
  //-------------------------------------------------------------------
  // Some async operation to get data, like a call to a 3-rd party API.
  //-------------------------------------------------------------------
  return 'retrieved async data, a string in this example';
}

async function numberLoader(): number {
  //-------------------------------------------------------------------
  // Some async operation to get data, like a call to a 3-rd party API.
  //-------------------------------------------------------------------
  return 123;
}


function SampleReactComponent() {
  // This works, and automatically deduces the correct `data` type as `string`.
  const { data, loading, timestamp } = useAsyncData('some.path', loader);

  // This gives TS error, as it able to evaluate the correct type of data
  // in the envelope at "some.path" is "string", not a "number" returned by
  // `numberLoader()`.
  const { data, loading, timestamp } = useAsyncData('some.path', numberLoader);

  // We can force a number, though, if we are sure about it.
  // NOTE: The "1" as the first generic parameter is just a "switch" value
  // enabling this overload of the hook, it has nothing to do with the "number"
  // type of data we are forcing here.
  const { data, loading, timer } = useAsyncData<ForceT, number>('some.path', numberLoader);

  return /* Some JSX markup. */
}
```

## Further Reading
- [Getting Started](/docs/tutorials/getting-started)
- (**Outdated**) [Blog Article](https://dr.pogodin.studio/dev-blog/the-global-state-in-react-designed-right)
  &zwnj; &mdash; _written in 2020 it covers the original motivation and features,
  but not the&nbsp;TypeScript flavour of the library._

## Frequently Asked Questions {#faq}

- _Does React Global State library avoid unnecessary component re-renders when values updated in the global state are irrelevant to those components?_

  Yes, it does avoid unnecessary re-renders of the component tree. A component
  relying on `some.path` in the global state is re-rendered only when the value
  at this path, or its sub-path has changed; _i.e._ it will be re-rendered if
  the value at `some.path` has changed, and it will be re-rendered if the value
  at `some.path.sub.path` has changed.

- _How would you describe your use case compared to another React global state library, e.g. [Valtio](https://www.npmjs.com/package/valtio)?_

  1.  React Global State is designed to follow the standard React API as close
      as possible. _E.g._ if some component relies on the local state:
      ```jsx
      const [value, setValue] = useState(initialState);
      ```
      to move that value to the global state (or _vice versa_) one only needs to
      replace the hook with
      ```jsx
      const [value, setValue] = useGlobalState(path, initialState);
      ```
      The [useGlobalState()] hook takes care to follow all edge cases of the
      standard [useState()]: `setValue` setter identity is stable (does not
      change on re-renders), functional updates and lazy initial state are
      supported.

      Other libraries tend to re-invent the wheel, introducing their own APIs,
      which (i) should be learned and understood; (ii) do complicate migration
      of components between the local and global state, should it be needed in
      a course of app development / prototyping.

  2.  When it comes to async data in the global state other libraries tend to
      offer only a very basic supported, often relying on experimental or internal
      React mechanics.

      React Global State, [useAsyncData()] and [useAsyncCollection()] hooks in
      particular, implements async data fetching and management features: when
      multiple components use these hooks to load async data to the same global
      state path the library takes care to do the actual loading just once, and
      then keep the data without reloading until their age reaches (configurable)
      max age. There is an automated garbage collection of expired, non-used
      async data from the global state; there is server-side rendering (SSR)
      support, with suggested high-level setup taking care that all async data
      loaded using [useAsyncData()] and [useAsyncCollection()] hooks will be
      automatically loaded and used in server-side renders (still allowing to
      opt-out of that for individual hooks, and timeout server-side fetching of
      data that take too long to arrive, in which case the library will fetch
      such data client-side). It does not rely on experimental React APIs to
      achieve its functionality, it only uses current public APIs.

      For me the support of async data fetching into the global state and their
      further management with out-of-the-box SSR support was the primary
      motivation to create React Global State. There are many other global state
      React libraries, but I was not able to find any that would cover the async
      data handling with that ease I believed was possible. The secondary
      motivation was that existing global state libraries either had
      the shortcoming of unnecessary component re-renders when data irrelevant
      to them where updated in the global state, or introduced their
      own APIs, where following the standard React APIs for local state looks
      to me a way more convenient approach.

- _Is React Global State library production ready (considering the current version number 0.y.z)?_

  Yes. I personally use it in production for all my commercial and personal
  React projects for over an year. I just don't feel like to call it v1 until
  a reasonable adoption by 3rd party developers, and any API improvements that
  may come out of community experience.

[Context API]: https://reactjs.org/docs/context.html
[functional updates]: https://reactjs.org/docs/hooks-reference.html#functional-updates
[Hooks]: https://reactjs.org/docs/hooks-intro.html
[lazy initial state]: https://reactjs.org/docs/hooks-reference.html#functional-updates
[React]: https://reactjs.orgs
[Redux]: https://redux.js.org
[useAsyncCollection()]: /docs/api/hooks/useasynccollection
[useAsyncData()]: /docs/api/hooks/useasyncdata
[useGlobalState()]: /docs/api/hooks/useglobalstate
[useState()]: https://reactjs.org/docs/hooks-reference.html#usestate
