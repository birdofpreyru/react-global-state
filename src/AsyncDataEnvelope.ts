export default class AsyncDataEnvelope<DataT> {
  data: null | DataT = null;

  numRefs: number = 0;

  operationId: string = '';

  timestamp: number = 0;
}
