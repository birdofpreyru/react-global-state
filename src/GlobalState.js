import _ from 'lodash';

/**
 * Transform state path into the full path inside GlobalState object.
 * @param {String} statePath
 */
function fullPath(statePath) {
  return _.isNil(statePath) ? 'state' : `state.${statePath}`;
}

/**
 * The GlobalState object powers data stores.
 */
export default class GlobalState {
  constructor(initialState) {
    this.state = _.cloneDeep(initialState);
    this.pendingNotification = false;
    this.watchers = [];
  }

  /**
   * Gets value from the `path`. If `path` is null or undefined, the entire
   * state object is returned.
   * @param {String} path
   * @return {Any}
   */
  get(path) {
    return _.get(this, fullPath(path));
  }

  /**
   * Sets `value` at the specified state `path`. If `path` is null or undefined,
   * the value is set as the entire state object.
   *
   * TODO: With such naive use of _.set, the state is mutated in place, which
   * may cause tons of unexpected side effects for dependants. It will be better
   * to partially clone the state, so that any existing references are not
   * mutated, while the full deep clonning is also avoided.
   *
   * @param {String} path
   * @param {Any} value
   * @return {Any} value
   */
  set(path, value) {
    _.set(this, fullPath(path), value);
    this.pendingNotification = true;
    setTimeout(() => {
      if (this.pendingNotification) {
        this.pendingNotification = false;
        this.watchers.forEach((w) => w());
      }
    });
    return value;
  }

  /**
   * Removes `callback` from watcher array. No operation if `callback` is not
   * in the watchers list.
   * @param {Function} callback
   */
  unWatch(callback) {
    const { watchers } = this;
    const pos = watchers.indexOf(callback);
    if (pos >= 0) {
      watchers[pos] = watchers[watchers.length - 1];
      watchers.pop();
    }
  }

  /**
   * Subscribes `callback` to state updates. No operation if `callback` is
   * in the watchers list already.
   * @param {Function} callback
   */
  watch(callback) {
    const { watchers } = this;
    if (watchers.indexOf(callback) < 0) {
      watchers.push(callback);
    }
  }
}
