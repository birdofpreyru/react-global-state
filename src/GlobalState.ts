import {
  get,
  isFunction,
  isObject,
  isNil,
  set,
  toPath,
} from 'lodash';

import SsrContext from './SsrContext';

import {
  type CallbackT,
  type ForceT,
  type LockT,
  type TypeLock,
  type ValueAtPathT,
  type ValueOrInitializerT,
  cloneDeepForLog,
  isDebugMode,
} from './utils';

const ERR_NO_SSR_WATCH = 'GlobalState must not be watched at server side';

type GetOptsT<T> = {
  initialState?: boolean;
  initialValue?: ValueOrInitializerT<T>;
};

export default class GlobalState<
  StateT,
  SsrContextT extends SsrContext<StateT> = SsrContext<StateT>,
> {
  readonly ssrContext?: SsrContextT;

  #dependencies: { [key: string]: Readonly<any[]> } = {};

  #initialState: StateT;

  // TODO: It is tempting to replace watchers here by
  // Emitter from @dr.pogodin/js-utils, but we need to clone
  // current watchers for emitting later, and this is not something
  // Emitter supports right now.
  #watchers: CallbackT[] = [];

  #nextNotifierId?: NodeJS.Timeout;

  #currentState: StateT;

  /**
   * Creates a new global state object.
   * @param initialState Intial global state content.
   * @param ssrContext Server-side rendering context.
   */
  constructor(
    initialState: StateT,
    ssrContext?: SsrContextT,
  ) {
    this.#currentState = initialState;
    this.#initialState = initialState;

    if (ssrContext) {
      /* eslint-disable no-param-reassign */
      ssrContext.dirty = false;
      ssrContext.pending = [];
      ssrContext.state = this.#currentState;
      /* eslint-enable no-param-reassign */

      this.ssrContext = ssrContext;
    }

    if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
      /* eslint-disable no-console */
      let msg = 'New ReactGlobalState created';
      if (ssrContext) msg += ' (SSR mode)';
      console.groupCollapsed(msg);
      console.log('Initial state:', cloneDeepForLog(initialState));
      console.groupEnd();
      /* eslint-enable no-console */
    }
  }

  /**
   * Drops the record of dependencies, if any, for the given path.
   */
  dropDependencies(path: string) {
    delete this.#dependencies[path];
  }

  /**
   * Checks if given `deps` are different from previously recorded ones for
   * the given `path`. If they are, `deps` are recorded as the new deps for
   * the `path`, and also the array is frozen, to prevent it from being
   * modified.
   */
  hasChangedDependencies(path: string, deps: any[]): boolean {
    const prevDeps = this.#dependencies[path];
    let changed = !prevDeps || prevDeps.length !== deps.length;
    for (let i = 0; !changed && i < deps.length; ++i) {
      changed = prevDeps![i] !== deps[i];
    }
    this.#dependencies[path] = Object.freeze(deps);
    return changed;
  }

  /**
   * Gets entire state, the same way as .get(null, opts) would do.
   * @param opts.initialState
   * @param opts.initialValue
   */
  getEntireState(opts?: GetOptsT<StateT>): StateT {
    let state = opts?.initialState ? this.#initialState : this.#currentState;
    if (state !== undefined || opts?.initialValue === undefined) return state;

    const iv = opts.initialValue;
    state = isFunction(iv) ? iv() : iv;
    if (this.#currentState === undefined) this.setEntireState(state);
    return state;
  }

  /**
   * Notifies all connected state watchers that a state update has happened.
   */
  private notifyStateUpdate(path: null | string | undefined, value: unknown) {
    if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
      /* eslint-disable no-console */
      const p = typeof path === 'string'
        ? `"${path}"` : 'none (entire state update)';
      console.groupCollapsed(`ReactGlobalState update. Path: ${p}`);
      console.log('New value:', cloneDeepForLog(value, path ?? ''));
      console.log('New state:', cloneDeepForLog(this.#currentState));
      console.groupEnd();
      /* eslint-enable no-console */
    }

    if (this.ssrContext) {
      this.ssrContext.dirty = true;
      this.ssrContext.state = this.#currentState;
    } else if (!this.#nextNotifierId) {
      this.#nextNotifierId = setTimeout(() => {
        this.#nextNotifierId = undefined;
        const watchers = [...this.#watchers];
        for (let i = 0; i < watchers.length; ++i) {
          watchers[i]!();
        }
      });
    }
  }

  /**
   * Sets entire state, the same way as .set(null, value) would do.
   * @param value
   */
  setEntireState(value: StateT): StateT {
    if (this.#currentState !== value) {
      this.#currentState = value;
      this.notifyStateUpdate(null, value);
    }
    return value;
  }

  /**
   * Gets current or initial value at the specified "path" of the global state.
   * @param path Dot-delimitered state path.
   * @param options Additional options.
   * @param options.initialState If "true" the value will be read
   *  from the initial state instead of the current one.
   * @param options.initialValue If the value read from the "path" is
   *  "undefined", this "initialValue" will be returned instead. In such case
   *  "initialValue" will also be written to the "path" of the current global
   *  state (no matter "initialState" flag), if "undefined" is stored there.
   * @return Retrieved value.
   */

  // .get() without arguments just falls back to .getEntireState().
  get(): StateT;

  // This variant attempts to automatically resolve and check the type of value
  // at the given path, as precise as the actual state and path types permit.
  // If the automatic path resolution is not possible, the ValueT fallsback
  // to `never` (or to `undefined` in some cases), effectively forbidding
  // to use this .get() variant.
  get<
    PathT extends null | string | undefined,
    ValueArgT extends ValueAtPathT<StateT, PathT, never>,
    ValueResT extends ValueAtPathT<StateT, PathT, void>,
  >(path: PathT, opts?: GetOptsT<ValueArgT>): ValueResT;

  // This variant is not callable by default (without generic arguments),
  // otherwise it allows to set the correct ValueT directly.
  get<Forced extends ForceT | LockT = LockT, ValueT = void>(
    path?: null | string,
    opts?: GetOptsT<TypeLock<Forced, never, ValueT>>,
  ): TypeLock<Forced, void, ValueT>;

  get<ValueT>(path?: null | string, opts?: GetOptsT<ValueT>): ValueT {
    if (isNil(path)) {
      const res = this.getEntireState((opts as unknown) as GetOptsT<StateT>);
      return (res as unknown) as ValueT;
    }

    const state = opts?.initialState ? this.#initialState : this.#currentState;

    let res = get(state, path);
    if (res !== undefined || opts?.initialValue === undefined) return res;

    const iv = opts.initialValue;
    res = isFunction(iv) ? iv() : iv;

    if (!opts?.initialState || this.get(path) === undefined) {
      this.set<ForceT, unknown>(path, res);
    }

    return res;
  }

  /**
   * Writes the `value` to given global state `path`.
   * @param path Dot-delimitered state path. If not given, entire
   * global state content is replaced by the `value`.
   * @param value The value.
   * @return Given `value` itself.
   */

  // This variant attempts automatic value type resolution & checking.
  set<
    PathT extends null | string | undefined,
    ValueArgT extends ValueAtPathT<StateT, PathT, never>,
    ValueResT extends ValueAtPathT<StateT, PathT, void>,
  >(path: PathT, value: ValueArgT): ValueResT;

  // This variant is disabled by default, otherwise allows to give
  // expected value type explicitly.
  set<Forced extends ForceT | LockT = LockT, ValueT = never>(
    path: null | string | undefined,
    value: TypeLock<Forced, never, ValueT>,
  ): TypeLock<Forced, void, ValueT>;

  set(path: null | string | undefined, value: unknown): unknown {
    if (isNil(path)) return this.setEntireState(value as StateT);

    if (value !== this.get(path)) {
      const root = { state: this.#currentState };
      let segIdx = 0;
      let pos: any = root;
      const pathSegments = toPath(`state.${path}`);
      for (; segIdx < pathSegments.length - 1; segIdx += 1) {
        const seg = pathSegments[segIdx]!;
        const next = pos[seg];
        if (Array.isArray(next)) pos[seg] = [...next];
        else if (isObject(next)) pos[seg] = { ...next };
        else {
          // We arrived to a state sub-segment, where the remaining part of
          // the update path does not exist yet. We rely on lodash's set()
          // function to create the remaining path, and set the value.
          set(pos, pathSegments.slice(segIdx), value);
          break;
        }
        pos = pos[seg];
      }

      if (segIdx === pathSegments.length - 1) {
        pos[pathSegments[segIdx]!] = value;
      }

      this.#currentState = root.state;

      this.notifyStateUpdate(path, value);
    }
    return value;
  }

  /**
   * Unsubscribes `callback` from watching state updates; no operation if
   * `callback` is not subscribed to the state updates.
   * @param callback
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  unWatch(callback: CallbackT) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    const pos = watchers.indexOf(callback);
    if (pos >= 0) {
      watchers[pos] = watchers[watchers.length - 1]!;
      watchers.pop();
    }
  }

  /**
   * Subscribes `callback` to watch state updates; no operation if
   * `callback` is already subscribed to this state instance.
   * @param callback It will be called without any arguments every
   * time the state content changes (note, howhever, separate state updates can
   * be applied to the state at once, and watching callbacks will be called once
   * after such bulk update).
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  watch(callback: CallbackT) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    if (watchers.indexOf(callback) < 0) {
      watchers.push(callback);
    }
  }
}
