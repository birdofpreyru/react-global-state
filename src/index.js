export { default as GlobalStateProvider } from './GlobalStateProvider';
export { default as useAsyncData } from './useAsyncData';
export { default as useGlobalState } from './useGlobalState';
