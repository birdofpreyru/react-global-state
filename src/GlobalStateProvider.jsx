/* eslint-disable react/prop-types */

import { createContext, useContext, useRef } from 'react';

import GlobalState from './GlobalState';

const context = createContext();

/**
 * Gets {@link GlobalState} instance from the context. In most cases
 * you should use {@link useGlobalState}, and other hooks to interact with
 * the global state, instead of accessing it directly.
 * @return {GlobalState}
 */
export function getGlobalState() {
  // Here Rules of Hooks are violated because "getGlobalState()" does not follow
  // convention that hook names should start with use... This is intentional in
  // our case, as getGlobalState() hook is intended for advance scenarious,
  // while the normal interaction with the global state should happen via
  // another hook, useGlobalState().
  /* eslint-disable react-hooks/rules-of-hooks */
  const globalState = useContext(context);
  /* eslint-enable react-hooks/rules-of-hooks */
  if (!globalState) throw new Error('Missing GlobalStateProvider');
  return globalState;
}

/**
 * @category Hooks
 * @desc Gets SSR context.
 * @param {boolean} [throwWithoutSsrContext=true] If `true` (default),
 * this hook will throw if no SSR context is attached to the global state;
 * set `false` to not throw in such case. In either case the hook will throw
 * if the {@link &lt;GlobalStateProvider&gt;} (hence the state) is missing.
 * @returns {SsrContext} SSR context.
 * @throws
 * - If current component has no parent {@link &lt;GlobalStateProvider&gt;}
 *   in the rendered React tree.
 * - If `throwWithoutSsrContext` is `true`, and there is no SSR context attached
 *   to the global state provided by {@link &lt;GlobalStateProvider&gt;}.
 */
export function getSsrContext(throwWithoutSsrContext = true) {
  const { ssrContext } = getGlobalState();
  if (!ssrContext && throwWithoutSsrContext) {
    throw new Error('No SSR context found');
  }
  return ssrContext;
}

/**
 * Provides global state to its children.
 * @prop {ReactNode} [children] Component children, which will be provided with
 * the global state, and rendered in place of the provider.
 * @prop {any} [initialState] Initial content of the global state.
 * @prop {SsrContext} [ssrContext] Server-side rendering (SSR) context.
 * @prop {boolean|GlobalState} [stateProxy] This option is useful for code
 * splitting and SSR implementation:
 * - If `true`, this provider instance will fetch and reuse the global state
 *   from a parent provider.
 * - If `GlobalState` instance, it will be used by this provider.
 * - If not given, a new `GlobalState` instance will be created and used.
 */
export default function GlobalStateProvider({
  children,
  initialState,
  ssrContext,
  stateProxy,
}) {
  const state = useRef();
  if (!state.current) {
    if (stateProxy instanceof GlobalState) state.current = stateProxy;
    else if (stateProxy) state.current = getGlobalState();
    else state.current = new GlobalState(initialState, ssrContext);
  }
  return (
    <context.Provider value={state.current}>
      {children}
    </context.Provider>
  );
}
