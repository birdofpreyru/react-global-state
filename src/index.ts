import GlobalState from './GlobalState';

import GlobalStateProvider, {
  getGlobalState,
  getSsrContext,
} from './GlobalStateProvider';

import SsrContext from './SsrContext';

import useAsyncCollection, {
  type UseAsyncCollectionI,
  type UseAsyncCollectionResT,
} from './useAsyncCollection';

import {
  type AsyncDataEnvelopeT,
  type AsyncDataLoaderT,
  type AsyncDataReloaderT,
  type UseAsyncDataI,
  type UseAsyncDataOptionsT,
  type UseAsyncDataResT,
  newAsyncDataEnvelope,
  useAsyncData,
} from './useAsyncData';

import useGlobalState, { type UseGlobalStateI } from './useGlobalState';

export type { AsyncCollectionLoaderT, AsyncCollectionT } from './useAsyncCollection';

export type { SetterT, UseGlobalStateResT } from './useGlobalState';

export type { ForceT, ValueOrInitializerT } from './utils';

export {
  type AsyncDataEnvelopeT,
  type AsyncDataLoaderT,
  type AsyncDataReloaderT,
  type UseAsyncCollectionResT,
  type UseAsyncDataOptionsT,
  type UseAsyncDataResT,
  getGlobalState,
  getSsrContext,
  GlobalState,
  GlobalStateProvider,
  newAsyncDataEnvelope,
  SsrContext,
  useAsyncCollection,
  useAsyncData,
  useGlobalState,
};

interface API<
  StateT,
  SsrContextT extends SsrContext<StateT> = SsrContext<StateT>,
> {
  getGlobalState: typeof getGlobalState<StateT, SsrContextT>;
  getSsrContext: typeof getSsrContext<SsrContextT>,
  GlobalState: typeof GlobalState<StateT, SsrContextT>,
  GlobalStateProvider: typeof GlobalStateProvider<StateT, SsrContextT>,
  newAsyncDataEnvelope: typeof newAsyncDataEnvelope,
  SsrContext: typeof SsrContext<StateT>,
  useAsyncCollection: UseAsyncCollectionI<StateT>,
  useAsyncData: UseAsyncDataI<StateT>,
  useGlobalState: UseGlobalStateI<StateT>,
}

const api = {
  getGlobalState,
  getSsrContext,
  GlobalState,
  GlobalStateProvider,
  newAsyncDataEnvelope,
  SsrContext,
  useAsyncCollection,
  useAsyncData,
  useGlobalState,
};

export function withGlobalStateType<
  StateT,
  SsrContextT extends SsrContext<StateT> = SsrContext<StateT>,
>() {
  return api as API<StateT, SsrContextT>;
}
