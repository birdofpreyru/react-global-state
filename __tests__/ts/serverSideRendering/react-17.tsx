import mockdate from 'mockdate';
import type { ComponentType } from 'react';
import ReactDOM from 'react-dom/server';

import {
  type MountedSceneT,
  consoleLogs,
  mockConsoleLog,
  unMockConsoleLog,
} from 'jest/utils';

import { GlobalStateProvider, SsrContext } from 'src/index';

jest.mock('uuid');

jest.useFakeTimers();
mockdate.set('2019-11-07Z');

let loaderA;
let loaderB;
let Scene: ComponentType;
let scene: MountedSceneT | undefined;
let ssrRound: number;

beforeEach(() => {
  ssrRound = 0;
  delete process.env.REACT_GLOBAL_STATE_DEBUG;
  unMockConsoleLog();
  ({ default: Scene, loaderA, loaderB } = require('./__assets__/TestScene'));
  loaderA.mockClear();
  loaderB.mockClear();
});

afterEach(() => {
  if (scene) {
    scene.destroy();
    scene = undefined;
  }

  // NOTE: Without this clearing some pending async updates from
  // "Scene test in the front-end mode" get fired only after the
  // later timer waiting in SSR test, which messes up the console
  // log snapshot verification.
  jest.clearAllTimers();
});

test('Naive server-side rendering', () => {
  const render = ReactDOM.renderToString((
    <GlobalStateProvider initialState={undefined}>
      <Scene />
    </GlobalStateProvider>
  ));
  expect(render).toMatchSnapshot();
});

/**
 * This is the sample SSR code assembly.
 */
async function serverSideRender(): Promise<string> {
  let render = '';
  ssrRound = 0;

  const ssrContext = new SsrContext({});

  for (; ssrRound < 10; ssrRound += 1) {
    /* eslint-disable no-await-in-loop */
    render = ReactDOM.renderToString((
      <GlobalStateProvider
        initialState={ssrContext.state}
        ssrContext={ssrContext}
      >
        <Scene />
      </GlobalStateProvider>
    ));
    if (ssrContext.dirty) {
      await Promise.allSettled(ssrContext.pending);
    } else break;
    /* eslint-disable no-await-in-loop */
  }
  return render;
}

test('Smart server-sider rendering', async () => {
  process.env.REACT_GLOBAL_STATE_DEBUG = '1';
  mockConsoleLog();
  const render = serverSideRender();
  await jest.runAllTimers();
  const renderString = await render;
  expect(ssrRound).toBe(1);
  expect(renderString).toMatchSnapshot();
  expect(consoleLogs).toMatchSnapshot();
});
