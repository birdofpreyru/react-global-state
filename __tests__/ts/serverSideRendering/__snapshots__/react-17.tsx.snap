// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Naive server-side rendering 1`] = `
"<div><h1>defaultValue</h1><div>{
  &quot;data&quot;: null,
  &quot;loading&quot;: false,
  &quot;timestamp&quot;: 0
}</div><div>{
  &quot;data&quot;: null,
  &quot;loading&quot;: false,
  &quot;timestamp&quot;: 0
}</div></div>"
`;

exports[`Smart server-sider rendering 1`] = `
"<div><h1>defaultValue</h1><div>{
  &quot;data&quot;: &quot;data&quot;,
  &quot;loading&quot;: false,
  &quot;timestamp&quot;: 1573084800000
}</div><div>{
  &quot;data&quot;: &quot;data&quot;,
  &quot;loading&quot;: false,
  &quot;timestamp&quot;: 1573084800000
}</div></div>"
`;

exports[`Smart server-sider rendering 2`] = `
[
  [
    "New ReactGlobalState created (SSR mode)",
  ],
  [
    "Initial state:",
    {},
  ],
  [
    "ReactGlobalState update. Path: "value.path"",
  ],
  [
    "New value:",
    "defaultValue",
  ],
  [
    "New state:",
    {
      "value": {
        "path": "defaultValue",
      },
    },
  ],
  [
    "ReactGlobalState update. Path: "x"",
  ],
  [
    "New value:",
    {
      "data": null,
      "numRefs": 0,
      "operationId": "",
      "timestamp": 0,
    },
  ],
  [
    "New state:",
    {
      "value": {
        "path": "defaultValue",
      },
      "x": {
        "data": null,
        "numRefs": 0,
        "operationId": "",
        "timestamp": 0,
      },
    },
  ],
  [
    "ReactGlobalState: async data (re-)loading. Path: "x"",
  ],
  [
    "ReactGlobalState update. Path: "x.operationId"",
  ],
  [
    "New value:",
    "S00000000-0000-0000-0000-000000000001",
  ],
  [
    "New state:",
    {
      "value": {
        "path": "defaultValue",
      },
      "x": {
        "data": null,
        "numRefs": 0,
        "operationId": "S00000000-0000-0000-0000-000000000001",
        "timestamp": 0,
      },
    },
  ],
  [
    "ReactGlobalState: async data (re-)loaded. Path: "x"",
  ],
  [
    "Data:",
    "data",
  ],
  [
    "ReactGlobalState update. Path: "x"",
  ],
  [
    "New value:",
    {
      "data": "data",
      "numRefs": 0,
      "operationId": "",
      "timestamp": 1573084800000,
    },
  ],
  [
    "New state:",
    {
      "value": {
        "path": "defaultValue",
      },
      "x": {
        "data": "data",
        "numRefs": 0,
        "operationId": "",
        "timestamp": 1573084800000,
      },
    },
  ],
  [
    "New ReactGlobalState created (SSR mode)",
  ],
  [
    "Initial state:",
    {
      "value": {
        "path": "defaultValue",
      },
      "x": {
        "data": "data",
        "numRefs": 0,
        "operationId": "",
        "timestamp": 1573084800000,
      },
    },
  ],
]
`;
